src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BulkScanSasTokenCache.java

```diff
@@ -0,0 +1,82 @@
+package uk.gov.hmcts.reform.blobrouter.services.storage;
+
+import com.azure.storage.common.implementation.Constants;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.blobrouter.clients.bulkscanprocessor.BulkScanProcessorClient;
+
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.TemporalAccessor;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static com.azure.storage.common.implementation.Constants.UrlConstants.SAS_EXPIRY_TIME;
+import static com.azure.storage.common.implementation.StorageImplUtils.parseQueryString;
+import static java.time.temporal.ChronoField.INSTANT_SECONDS;
+
+@Service
+public class BulkScanSasTokenCache {
+
+    private final BulkScanProcessorClient bulkScanSasTokenClient;
+    private final long refreshSasBeforeExpiry;
+
+    private static Cache<String, String> tokenCache;
+
+    public BulkScanSasTokenCache(
+        BulkScanProcessorClient bulkScanSasTokenClient,
+        @Value("${bulk-scan-cache.refresh-before-expire-in-sec}") long refreshSasBeforeExpiry
+    ) {
+        this.bulkScanSasTokenClient = bulkScanSasTokenClient;
+        this.refreshSasBeforeExpiry = refreshSasBeforeExpiry;
+        tokenCache = Caffeine.newBuilder()
+            .expireAfter(new BulkScanContainerCacheExpiry())
```

Comment:
can we not just do
```
Caffeine.newBuilder()
  .expireAfterWrite(4, TimeUnit.MINUTES)

```
so cache expires after every 4 minutes ?

Reply:
we can but it is coupling the processor and blob router.
processor can change it anytime?

Reply:
I think we should be fine given that it has not changed from last 1 year. Also if it. changes then we can easily update the config.
Worth checking with other devs

Reply:
I can not see any harm to use expiry info which is already available.

Reply:
well the code is not that clear to understand imo

Reply:
I'm fine with both. I like the fact that it takes the expiry time from SAS token.